server/src/main/java/com/cloud/hypervisor/KVMGuru.java

```diff
@@ -297,4 +309,49 @@ public Map<String, String> getClusterSettings(long vmId) {
         return null;
     }
 
-}
+    @Override
+    public VirtualMachine importVirtualMachineFromBackup(long zoneId, long domainId, long accountId, long userId, String vmInternalName, Backup backup) throws Exception {
+        s_logger.debug(String.format("Trying to import VM [vmInternalName: %s] from Backup [%s].", vmInternalName,
+                ReflectionToStringBuilderUtils.reflectOnlySelectedFields(backup, "id", "uuid", "vmId", "externalId", "backupType")));
+
+        VMInstanceVO vm = _instanceDao.findVMByInstanceNameIncludingRemoved(vmInternalName);
+        if (vm.getRemoved() != null) {
+            vm.setState(VirtualMachine.State.Stopped);
+            vm.setPowerState(VirtualMachine.PowerState.PowerOff);
+            _instanceDao.update(vm.getId(), vm);
+            _instanceDao.unremove(vm.getId());
+        }
+        for (final VolumeVO volume : _volumeDao.findIncludingRemovedByInstanceAndType(vm.getId(), null)) {
+            volume.setState(Volume.State.Ready);
+            volume.setAttached(new Date());
+            _volumeDao.update(volume.getId(), volume);
+            _volumeDao.unremove(volume.getId());
+        }
+
+        return vm;
+    }
+
+
+
+    @Override public boolean attachRestoredVolumeToVirtualMachine(long zoneId, String location, Backup.VolumeInfo volumeInfo, VirtualMachine vm, long poolId, Backup backup)
+            throws Exception {
+
+        VMInstanceVO targetVM = _instanceDao.findVMByInstanceNameIncludingRemoved(vm.getName());
+        List<VolumeVO> devices = _volumeDao.findIncludingRemovedByInstanceAndType(targetVM.getId(), null);
+        VolumeVO restoredVolume = _volumeDao.findByUuid(location);
+        Integer deviceId = devices.size();
+
+
+        if (restoredVolume != null) {
+            restoredVolume.setState(Volume.State.Ready);
+            _volumeDao.update(restoredVolume.getId(), restoredVolume);
+            try {
+                _volumeDao.attachVolume(restoredVolume.getId(), vm.getId(), deviceId);
```

Comment:
Hi, @fermosan, I'm not sure, but I think you will have a problem if you have a VM with more volumes and restore them at some point. They will have the same `deviceId`. The issue will appear at the start of the VM when the `LibvirtComputingResource::createVbd` is invoked.

Reply:
Hello @slavkap. We can test this if you think of a specific test case.  attachRestoredVolumeToVirtualMachine() is supposed to be called for only one volume at a time from B&R framework (attachVolumeToVM in class BackupManagerImpl). 
Each time it is called and depending on the currently  attached devices  for that vm it will return the next counter.

Reply:
@slavkap it should now be settled. Thanks

Reply:
Thanks, @fermosan, this should work! You can extract the repeated code from `importVirtualMachineFromBackup` and `attachRestoredVolumeToVirtualMachine` into a method `getDeviceId` or something

Reply:
Done